### ISSUE TITLE
A question about the mean value of the iterative sampling

### ISSUE THREAD

**Issue by YihanLi126** [sisl/MPOPIS#8]
Hello!

I have another question about the code here:
https://github.com/sisl/MPOPIS/blob/ceff32bdc81cfb4b00e2115d40447549c174647f/src/mppi_mpopi_policies.jl#L465
In the code above, the mean value of the sampling is added to `pol.U` in every iteration, and in the following code, the difference between `pol.U` and its original value is added to the final sampling result `E`(LHS):
https://github.com/sisl/MPOPIS/blob/ceff32bdc81cfb4b00e2115d40447549c174647f/src/mppi_mpopi_policies.jl#L468
I think what it means is, the mean value of the final result of the sampling `E`(LHS) is the sum of the mean value in every iteration. 

Can this cause the predictive trajectory calculated from `E` biases from the sampling in every iteration? Is it reasonable to just use the mean value of the elite trajectory in the last iteration as the mean value of the final sampling result instead of adding all those mean values up?

Thank you for your patience!

**Comment by dylan-asmar** [comment]
That section is for the cross-entropy (CE) method integrated with MPOPI. For the CE method, we perform sampling from some distribution, calculate the costs associated with those samples, select "elite" samples, and then use those "elite" samples to update our sampling distribution. We would repeat this process when using the CE method for pure optimization until some stopping criteria. (Section 8.4 of [this book](https://algorithmsbook.com/optimization/) is a good reference).

To use CE as adaptive importance sampling within MPPI, we first sample from the proposal distribution which would have the mean as our nominal control (rolled over from our last time step), and in this implementation, we just set the initial covariance based on some hyperparameter (normal considerations with CE on selecting this value apply, e.g. large enough to not get stuck in local optima while small enough to allow convergence after a certain number of iterations). Then we need to update the mean and covariance at each iteration. When we run out of iterations, we take the new mean of our distribution as the control policy.

The way I implemented it in code is a bit confusing (sorry about that), but follows the same logic. We sample from a Normal distribution with a mean of 0. So a way to think about this distribution is the "change" in control from our nominal.
https://github.com/sisl/MPOPIS/blob/ceff32bdc81cfb4b00e2115d40447549c174647f/src/mppi_mpopi_policies.jl#L447
The sampled change in control is added to the nominal control in the `simulate_model` function: https://github.com/sisl/MPOPIS/blob/ceff32bdc81cfb4b00e2115d40447549c174647f/src/mppi_mpopi_policies.jl#L271
Line 465 (`pol.U = pol.U + vec(mean(elite, dims=2))` is updating the "mean" of our proposal distribution. We still sample from a Normal with a mean of 0, but we shift our nominal control at each iteration of the CE method. I decided on this implementation choice to allow for an easier way to track control cost.

Line 468 is shifting the 0 mean samples to the "new" mean (our updated nominal control) before we go back to determine our final control:
https://github.com/sisl/MPOPIS/blob/ceff32bdc81cfb4b00e2115d40447549c174647f/src/mppi_mpopi_policies.jl#L221

Let me know if this answers your questions or if you are still uncertain.
